### PR TITLE
DEV: Refactor publish_post_after_save function

### PR DESCRIPTION
**Pull request by scossar**
Refactors the `publish_post_after_save` function. It's logic has been broken out into the following protected functions:

- `exclude_post`
- `auto_publish`
- `force_publish_post`

I've included the new protected functions directly under the `publish_post_after_save` function. This was done for readability reasons, but possibly they should be moved lower down in the file.

Some variables have been renamed in the attempt to make it clear at first glance what they are being used for.

None of the underlying logic has been changed.

This PR also removes the unused `post_id` argument from the `has_excluded_tag` function and fixes an issue where `$post_id` was being passed instead of `$post` to that function from the `xmlrpc_publish_post_to_discourse` function.

Possibly some similar changes should be made to the `sync_to_discourse_work` function.

### CODE REVIEW

**Comment by angusmcleod** [comment]
@scossar Thanks! I'll take a look at this closely on monday.

File: lib/discourse-publish.php
Comment (angusmcleod) [inline lib/discourse-publish.php:173]: @scossar To keep this method focused on a single boolean outcome (i.e. force_publish) perhaps we should move this post meta update outside into the `publish_post_after_save` method, conditional on force_publish?

**Review by angusmcleod — CHANGES_REQUESTED** [review]
@scossar Thanks for this, this is a good direction to go in. Once we've merged this I think I'll add some more unit tests in a separate PR.

**Comment by scossar** [comment]
Both those changes make sense to me. I'll update the PR tomorrow morning.

**Comment by scossar** [comment]
With the call to `update_post_meta` moving out of the `force_publish_post` function, I guess it would make sense to apply the same logic to its return statement as to the `exclude_post` return statement. Something like:

```
return ( ( 0 === $force_publish_max_age ) || $post_time >= $min_date ) && $is_rest_request;
```

**Comment by scossar** [comment]
I've made the changes.

Arguably, the `$force_publish_post` variable could be set before the `$publish_to_discourse` variable. Then `$publish_to_discourse` could be set with:

```
$publish_to_discourse = $publish_new_post_to_discourse || $topic_should_be_updated || $force_publish_post;
```

**Comment by angusmcleod** [comment]
> Arguably, the `$force_publish_post` variable could be set before the `$publish_to_discourse` variable. Then `$publish_to_discourse` could be set with:
> 
> ```
> $publish_to_discourse = $publish_new_post_to_discourse || $topic_should_be_updated || $force_publish_post;
> ```

Yes, would you mind having a look at this change in this PR as well? We may as well make that move at the same time here, as part of the same package of changes. Query whether this will require a new unit test?

**Comment by scossar** [comment]
> Query whether this will require a new unit test?

Couldn't hurt. It's a setting that could cause things to go very wrong. 

I've got the test suite running locally now. I had to downgrade to PHP 7.4 to install the packages. If you can hold off until tomorrow, I'll make the change I suggested above and see about adding a new unit test.

File: tests/phpunit/test-discourse-publish.php
Comment (scossar) [inline tests/phpunit/test-discourse-publish.php:36]: Not sure if there's a better way to do this.

**Comment by scossar** [comment]
I've updated the PR and added some tests for the 'force-publish' option. I got stumped on how to deal with this condition: https://github.com/discourse/wp-discourse/blob/main/lib/discourse-publish.php#L109. It's there to make sure that 'force-publish' is only applied to posts published via the block editor. That satisfied the need of the client who wanted the option, but it seems it would be better to allow the option to function with both the block and classic editors.

The test is currently defining `REST_REQUEST` and setting it to `true` in its `setUpBeforeClass` method. I don't think that will cause any unintended side effects, but it seems wrong.

**Comment by angusmcleod** [comment]
@scossar Thanks for that further explanation. To me this is a sign we need to further abstract the code first, both to make it easier to test, and to make it easier to support additional options. I'm thinking we do the following

1. Move `$is_rest_request` in `force_publish_post` to a separate method, e.g. `force_publish_allowed`
2. Make the new method (`force_publish_allowed`) able to be overridden in a unit test (e.g. using class variable)
3. Move the use of `force_publish_allowed` to the guard conditional in `force_publish_post`.

That will:
- Allow us to more easily test this (without setting env variables).
- Make it easier to support the classic editor if we so desire.
- Clarify the code further

**Comment by scossar** [comment]
That sounds like a good approach to me. I think I've got it.

To avoid adding a lot of of tests, I updated the tests to handle the case of a post not being published and a post being published within a single method. Also to ensure that the "doesn't work" case doesn't pass for the wrong reason. I'm not seeing any potential side effects from doing this, but don't mind breaking the two cases into separate methods if that's better.

Also, I ran `phpcbf --standard=WordPress` for `test-discourse-publish.php`. Unfortunately that generated a huge number of formatting changes. Never sure how to deal with that in a PR.

**Comment by angusmcleod** [comment]
> To avoid adding a lot of of tests, I updated the tests to handle the case of a post not being published and a post being published within a single method. Also to ensure that the "doesn't work" case doesn't pass for the wrong reason. I'm not seeing any potential side effects from doing this, but don't mind breaking the two cases into separate methods if that's better.

I'm happy with this 👍 

> Also, I ran phpcbf --standard=WordPress for test-discourse-publish.php. Unfortunately that generated a huge number of formatting changes. Never sure how to deal with that in a PR.

We now use the specific phpcbf settings in `.phpcs.xml` as explained in [FORMATTING.md](https://github.com/discourse/wp-discourse/blob/main/docs/FORMATTING.md), which your PR is passing (it's in the CI), so you're good on that front.

Is there anything else you wanted to tweak before I merge this? It won't be automatically pushed as part of a version, but it'll be part of the version I push next week.

**Comment by scossar** [comment]
>Is there anything else you wanted to tweak before I merge this?

The `xmlrpc_publish_post` method (https://github.com/discourse/wp-discourse/blob/main/lib/discourse-publish.php#L143) could probably use looking at:

There should probably be a test that ensures it doesn't call `sync_to_discourse` unless a function that hooks into `wp_discourse_before_xmlrpc_publish` has been added to a theme or plugin. 

It probably makes sense to keep it as a separate method that's hooked into from the `xmlrpc_publish_post` action, but maybe that method could call `publish_post_after_save( $post_id, $post )` instead of duplicating its code. (Not sure about this)

Posts published via `xmlrpc` (blogging software) have the same issue with a Discourse category not being set as force-published posts do. Looking at that code now, I'm not sure that the category for force-published posts needs to be set in the `publish_post_after_save` method. The `sync_to_discourse_work` method handles the case of a category not being set: https://github.com/discourse/wp-discourse/blob/main/lib/discourse-publish.php#L267. 

Related to the above, sanitizing the title in `publish_post_after_save` feels a little off. That method should just be used to determine if `sync_to_discourse` gets called for a post.

I'd be ok with the PR getting merged as it is now and making a new PR that looks into the above issues. Whatever works best for you in terms of timing. I can probably get to the issues I've mentioned here before your monday.

**Comment by angusmcleod** [comment]
> I'd be ok with the PR getting merged as it is now and making a new PR that looks into the above issues.

Yes, I think that'd be best.

> Whatever works best for you in terms of timing. I can probably get to the issues I've mentioned here before your monday.

That's fine. Thanks again!

**Review by angusmcleod — APPROVED** [review]

**Review by Jailyard90Grad — APPROVED** [review]

**Pull request merged**
